tempesta_fw/tls.c

```diff
@@ -19,73 +19,583 @@
  * this program; if not, write to the Free Software Foundation, Inc., 59
  * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
  */
+#include "cfg.h"
 #include "connection.h"
+#include "client.h"
 #include "tls.h"
 
+#define tls_dbg(c, fmt, ...)					\
+	TFW_DBG("TLS(%p,%p):%s " fmt "\n", c, c->sk, __func__, ## __VA_ARGS__)
+#define tls_err(c, fmt, ...)					\
+	TFW_ERR("TLS(%p,%p):%s " fmt "\n", c, c->sk, __func__, ## __VA_ARGS__)
+
+#define MAX_TLS_PAYLOAD		1460
+
+typedef struct {
+	mbedtls_ssl_config	cfg;
+	mbedtls_x509_crt	crt;
+	mbedtls_pk_context	key;
+} TfwTls;
+
+static TfwTls tfw_tls = { 0 };
+
+/* Cache for fast TfwTlsContext{} allocations. */
+static struct kmem_cache *tfw_tls_cache = NULL;
+
+static inline int
+ttls_ssl_handshake(TfwTlsContext *tls)
+{
+	int r;
+
+	spin_lock(&tls->lock);
+	r = mbedtls_ssl_handshake(&tls->ssl);
+	spin_unlock(&tls->lock);
+
+	return r;
+}
+
+static inline int
+ttls_ssl_read(TfwTlsContext *tls, unsigned char *data, size_t size)
+{
+	int r;
+
+	spin_lock(&tls->lock);
+	r = mbedtls_ssl_read(&tls->ssl, data, size);
+	spin_unlock(&tls->lock);
+
+	return r;
+}
+
+static inline int
+ttls_ssl_write(TfwTlsContext *tls, const unsigned char *data, size_t size)
+{
+	int r;
+
+	spin_lock(&tls->lock);
+	r = mbedtls_ssl_write(&tls->ssl, data, size);
+	spin_unlock(&tls->lock);
+
+	return r;
+}
+
+static inline int
+ttls_ssl_close_notify(TfwTlsContext *tls)
+{
+	int r;
+
+	spin_lock(&tls->lock);
+	r = mbedtls_ssl_close_notify(&tls->ssl);
+	spin_unlock(&tls->lock);
+
+	return r;
+}
+
 /**
- * TODO do all crypto and handle TLS FSM here.
- *
  * Decrypted response messages should be directly placed in TDB area
  * to avoid copying.
  */
 static int
 tfw_tls_msg_process(void *conn, struct sk_buff *skb, unsigned int off)
 {
-	int r = TFW_BLOCK;
-	TfwConnection *c = (TfwConnection *)conn;
-	TfwMsg *msg = (TfwMsg *)c->msg;
+	int r;
+	TfwConnection *c = conn;
+	TfwTlsContext *tls = c->tls;
 
-	/*
-	 * TODO switch to HTTP FSM, @data and @len must be decrypted here.
-	 * Typically we don't need original skb any more either for server or
-	 * proxy modes, so it has sense to do decryption in-place.
-	 */
-	r = tfw_gfsm_move(&msg->state, TFW_HTTPS_FSM_TODO_ISSUE_81, skb, off);
+	tls_dbg(c, "=>");
 
-	return r;
+	ss_skb_queue_tail(&tls->rx_queue, skb);
+
+	r = ttls_ssl_handshake(tls);
+	if (r == MBEDTLS_ERR_SSL_CONN_EOF) {
+		return TFW_PASS; /* more data needed */
+	} else if (r == 0) {
+		struct sk_buff *nskb;
+
+		nskb = alloc_skb(MAX_TCP_HEADER + MAX_TLS_PAYLOAD, GFP_ATOMIC);
+		if (unlikely(!nskb))
+			return TFW_BLOCK;
+
+		skb_reserve(nskb, MAX_TCP_HEADER);
+		skb_put(nskb, MAX_TLS_PAYLOAD);
+
+		r = ttls_ssl_read(tls, nskb->data, MAX_TLS_PAYLOAD);
+		if (r == MBEDTLS_ERR_SSL_WANT_READ ||
+		    r == MBEDTLS_ERR_SSL_WANT_WRITE) {
```

Comment:
Please place `{` at new line for multi-line IF statements - that's more readable.